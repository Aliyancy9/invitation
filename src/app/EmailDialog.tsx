import * as React from 'react'
import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import { makeStyles } from '@material-ui/styles'
import registerEmail from './registerEmail'
import { Formik, FormikHelpers } from 'formik'
import * as Yup from 'yup'
import { Typography, Divider } from '@material-ui/core'

const useStyles = makeStyles({
  action: {
    margin: '1rem 0',
    padding: 0,
    justifyContent: 'center',
  },
  title: {
    display: 'flex',
    justifyContent: 'center',
    margin: '2rem',
    fontStyle: 'oblique',
    color: 'grey',
  },
  form: {
    display: 'flex', 
    flexDirection:'column',
    '& > *:not(:last-child)': {
      height: '5rem',
    },
  },
  input: {
    height: '0.8rem',
  },
  message: {
    display: 'flex',
    justifyContent: 'center',
  },
  successButton: {
    margin: '2rem 6rem'
  }, 
})

export interface Email {
  name: string
  email: string
  confirmedEmail?: string
} 

export interface Props {
    isOpen: boolean
    onClose: () => void
}

const EmailDialog: React.FC<Props> = (
  {
    isOpen,
    onClose
  }) => {  
    
  const classes = useStyles({})
  const initialValues: Email = {
    name: '',
    email: '',
    confirmedEmail: ''
}

  const validationSchema = Yup.object().shape({
  name: Yup.string()
    .min(3)
    .required('Required'),

  email: Yup.string()
    .email()
    .required('Required'),
  
  confirmedEmail: Yup.string()
    .required('Required')
    .oneOf([Yup.ref('email'), null], 'Email must match')
})

  const [isSubmitionCompleted, setSubmitionCompleted] = React.useState<Boolean>(false)
  const [hasErrorMessage, setHasErrorMessage] = React.useState<Boolean>(false)
  const [errorMessage, setErrorMessage] = React.useState('')
    
  const handleOnSubmit = async(email: Email, actions: FormikHelpers<Email>) => {
    setErrorMessage('')
    const emailToSubmit = {
      name: email.name,
      email: email.email
    }
   
    const res = await registerEmail(emailToSubmit)
    actions.setSubmitting(false)
    
    if(res.errorMessage) {
      setHasErrorMessage(true)
      setErrorMessage(res.errorMessage)
    } else {
      setSubmitionCompleted(true)
    }
  }

  const handleOnClose = async () => {
    onClose()
    await new Promise(res => setTimeout(res, 500))
    setSubmitionCompleted(false) 
  }

    return ( 
      <Dialog
        open={isOpen}
        fullWidth
        maxWidth= {isSubmitionCompleted? 'sm': 'xs'}
      >
        {!isSubmitionCompleted && ( 
          <React.Fragment>
          <DialogTitle className={classes.title}> 
              Request an invite
              <Divider variant='middle'/>
          </DialogTitle>
          <DialogContent>
            <Formik
              initialValues={initialValues}
              onSubmit={handleOnSubmit}
              validateOnChange={false}
              validateOnBlur={false}
              validationSchema={validationSchema}
            >
              {(props) => {
                const {
                  values,
                  errors,
                  isSubmitting,
                  handleChange,
                  handleSubmit,
                  isValid,
                } = props;
                return (
                  <form onSubmit={handleSubmit} className={classes.form}>
                    <TextField
                      data-testid='name-input'
                      label='Full name'
                      name='name'
                      variant='outlined'
                      error={!!errors.name}
                      InputProps={{ classes: { input: classes.input } }}
                      value={values.name}
                      onChange={handleChange}
                      helperText={errors.name}           
                      disabled={isSubmitting}
                    />

                    <TextField
                      data-testid='email-input'
                      InputProps={{ classes: { input: classes.input } }}
                      error={!!errors.email }
                      label='Email'
                      name='email'
                      variant='outlined'
                      value={values.email}
                      onChange={handleChange}
                      helperText={errors.email}      
                      disabled={isSubmitting}
                    />
                    <TextField
                      data-testid='confirmedEmail-input'
                      InputProps={{ classes: { input: classes.input } }}
                      error={!!errors.confirmedEmail}
                      label='Confirm email'
                      name='confirmedEmail'
                      variant='outlined' 
                      value={values.confirmedEmail}
                      onChange={handleChange}
                      helperText={errors.confirmedEmail}
                      disabled={isSubmitting}
                    />
                    <DialogActions  
                      classes={{
                        root: classes.action,
                      }}
                      >
                      <Button type='submit'  disabled={isSubmitting} fullWidth color='default' variant='outlined'>
                        {isSubmitting? 'Sendinging please wait...': 'Send'}
                      </Button>
                    </DialogActions>
                    {isValid && hasErrorMessage && (
                    <Typography color = 'textSecondary' className={classes.message}>
                      {errorMessage}
                    </Typography>
                   )}    
                  </form>
                )
              }}
            </Formik>
          </DialogContent>
          </React.Fragment>
          )} 
          {isSubmitionCompleted && (
            <React.Fragment>
              <DialogTitle className={classes.title}> 
                  All done!
                  <Divider variant='middle'/>
              </DialogTitle>
                <Typography variant = 'h6' color = 'textSecondary' className={classes.message}>
                  You will be the one of the first to experience
                </Typography>
                <Typography variant = 'h6' color = 'textSecondary' className={classes.message}>
                  Broccoli & Co. when we launch
                </Typography>
              <DialogActions  
                classes={{
                  root: classes.successButton,
                }}
                >
                <Button fullWidth color='default' variant='outlined' onClick={handleOnClose}>
                  ok
                </Button>
              </DialogActions>
            </React.Fragment>
          )}   
      </Dialog>
  )
}

export default EmailDialog