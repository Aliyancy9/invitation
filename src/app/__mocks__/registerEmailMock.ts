const registerEmailMock = jest.fn(() => Promise.resolve('Registeted')) as jest.Mock<any>

export default registerEmailMock
