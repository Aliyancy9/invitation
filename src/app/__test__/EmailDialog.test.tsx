import * as React from 'react'
import { render, waitForElement, fireEvent,act } from '@testing-library/react'

import registerEmailMock from '../__mocks__/registerEmailMock'

import EmailDialog from '../EmailDialog'

const baseProps = {
    isOpen: true,
    onClose: jest.fn()
}

const renderTest = (comp: any) => render(comp)

describe('EmailDialog', () => {
    it('should render correctly', () => {
         const { container } = renderTest(<EmailDialog {...baseProps} />)
         expect(container).toMatchSnapshot()

    })
    it('should show correct title', async () => {
    const { getByText } = renderTest(<EmailDialog {...baseProps} />)

        const titleElement = await waitForElement(() => getByText('Request an invite'))

        expect(titleElement.innerHTML).toContain('Request an invite')
    })

   
    it('should show correct input labels', async () => {
    const { getAllByText } = renderTest(<EmailDialog {...baseProps} />)
    const nameLabel = 'Full name'
    const emailLabel = 'Email'
    const confirmedEmailLabel = 'Confirm email'
   
    const nameElement = await waitForElement(() => getAllByText(nameLabel))
    const emailElement = await waitForElement(() => getAllByText(emailLabel))
    const confirmedEmailElement = await waitForElement(() => getAllByText(confirmedEmailLabel))

    expect(nameElement).toBeTruthy()
    expect(emailElement).toBeTruthy()
    expect(confirmedEmailElement).toBeTruthy()

})

    it('should only Submit the form if the values are correct', async () => {
    const { getByText, getByTestId } = renderTest(<EmailDialog {...baseProps} />)
    
    const nameInput = getByTestId('name-input').querySelector('input')
    const emailInput = getByTestId('email-input').querySelector('input')
    const confirmedEmailInput =getByTestId('confirmedEmail-input').querySelector('input')
    
    act(() => {fireEvent.change(nameInput, { target: { value: 'doris' }})})
    act(() => {fireEvent.change(emailInput, { target: { value: 'email@example.com' }})})
    act(() => {fireEvent.change(confirmedEmailInput, { target: { value: 'email@example.com' } })})
    
    const addButton = getByText('Send')
    fireEvent.click(addButton)

    expect(registerEmailMock.mockReturnValue('Registered'))
    })
})