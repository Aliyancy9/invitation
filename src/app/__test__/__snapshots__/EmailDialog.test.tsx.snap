// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailDialog should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
